src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+
+            public CompilationTaskNode(CompilationTask task)
+            {
+                Task = task;
+                Children = new Dictionary<string, CompilationTaskNode>();
+            }
+        }
+
+        /// <summary>
+        /// Represents a warning type detected while tracking compilation events.
+        /// </summary>
+        private enum WarningType
+        {
+            ProcessAlreadyExists,
+            ProcessDoesNotExist,
+            ProcessAlreadyEnded
+        }
+
+        /// <summary>
+        /// Represents a warning detected while tracking compilation events.
+        /// </summary>
+        private class Warning
+        {
+            public readonly WarningType Type;
+            public readonly DateTime UtcDateTime;
+            public readonly string Key;
+
+            public Warning(WarningType type, string key)
+            {
+                UtcDateTime = DateTime.UtcNow;
+                Type = type;
+                Key = key;
+            }
+        }
+
+        // Private members.
+
+        /// <summary>
+        /// Defines a handler for a type of compilation task event.
+        /// </summary>
+        private delegate void CompilationTaskEventTypeHandler(CompilationLoader.CompilationTaskEventArgs eventArgs);
+
+        /// <summary>
+        /// Contains a handler that takes care of each type of task event.
+        /// </summary>
+        private static readonly IDictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler> CompilationEventTypeHandlers = new Dictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler>
+        {
+            { CompilationLoader.CompilationTaskEventType.Start, CompilationEventStartHandler },
+            { CompilationLoader.CompilationTaskEventType.End, CompilationEventEndHandler }
+        };
+
+        /// <summary>
+        /// Contains the compilation tasks tracked through the handled events.
+        /// </summary>
+        private static readonly IDictionary<string, CompilationTask> CompilationTasks = new Dictionary<string, CompilationTask>();
+        /// <summary>
+        /// Contains the warnings generated while handling the compiler tasks events.
+        /// </summary>
+        private static readonly IList<Warning> Warnings = new List<Warning>();
+
+        // Private methods.
+
+        /// <summary>
+        /// Creates a hierarchical structure that contains the compilation tasks.
+        /// </summary>
+        private static IList<CompilationTaskNode> BuildCompilationTasksHierarchy()
+        {
+            IList<CompilationTaskNode> compilationTasksForest = new List<CompilationTaskNode>();
+            Queue<CompilationTaskNode> toFindChildrenNodes = new Queue<CompilationTaskNode>();
+
+            // First add the roots (top-level tasks) of all trees to the forest.
+
+            foreach (KeyValuePair<string, CompilationTask> entry in CompilationTasks)
```

Comment:
The 'var' keyword is typically used for the iteration variable of a foreach loop.
```suggestion
            foreach (var entry in CompilationTasks)
```